File: Data Structures & Algorithms/linked-list-cycle-detection/submission-14.js
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     constructor(val = 0, next = null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

class Solution {
    /**
     * @param {ListNode} head
     * @return {boolean}
     */
    hasCycle(head) {
        if(head === null || head.next === null) return false;
        let slow = head, fast = head.next;

        while(fast.next !== null && fast.next.next !== null){
            if(slow === fast) return true;
            slow = slow.next;
            fast = fast.next.next;
        }

        return false;
    }
}

File: Data Structures & Algorithms/find-duplicate-integer/submission-14.py
class Solution:
    def findDuplicate(self, nums: List[int]) -> int:
        fast = 0
        slow = 0

        while True:
            slow = nums[slow]
            fast = nums[nums[fast]]

            if slow == fast:
                slow = 0
                while slow != fast:
                    slow = nums[slow]
                    fast = nums[fast]
                return slow
            
